src/frontend/src/scheduler/hummock_snapshot_manager.rs

```diff
@@ -13,125 +13,215 @@
 use std::sync::Arc;
+use std::time::{Duration, Instant};
 
+use anyhow::anyhow;
 use log::error;
-use tokio::sync::Mutex;
+use tokio::sync::mpsc::{channel, Sender};
+use tokio::sync::oneshot::{channel as once_channel, Sender as Callback};
 
 use crate::meta_client::FrontendMetaClient;
 use crate::scheduler::plan_fragmenter::QueryId;
-use crate::scheduler::SchedulerError::PinSnapshot;
-use crate::scheduler::SchedulerResult;
+use crate::scheduler::{SchedulerError, SchedulerResult};
+
+const MAX_WAIT_EPOCH_REQUEST_NUM: usize = 4096;
+const UNPIN_INTERVAL_SECS: u64 = 10;
 
 /// Cache of hummock snapshot in meta.
 pub struct HummockSnapshotManager {
-    core: Mutex<HummockSnapshotManagerCore>,
-    meta_client: Arc<dyn FrontendMetaClient>,
+    sender: Sender<EpochOperation>,
 }
 pub type HummockSnapshotManagerRef = Arc<HummockSnapshotManager>;
 
+enum EpochOperation {
+    RequestEpoch {
+        query_id: QueryId,
+        sender: Callback<SchedulerResult<u64>>,
+    },
+    ReleaseEpoch {
+        query_id: QueryId,
+        epoch: u64,
+    },
+    Tick,
+}
+
 impl HummockSnapshotManager {
     pub fn new(meta_client: Arc<dyn FrontendMetaClient>) -> Self {
-        Self {
-            core: Mutex::new(HummockSnapshotManagerCore::new()),
-            meta_client,
-        }
-    }
-
-    pub async fn get_epoch(&self, query_id: QueryId) -> SchedulerResult<u64> {
-        let mut core_guard = self.core.lock().await;
-        if core_guard.is_outdated {
-            let epoch = self
-                .meta_client
-                .pin_snapshot(core_guard.last_pinned)
-                .await
-                .map_err(|_e| PinSnapshot(query_id.clone(), core_guard.last_pinned))?;
-            core_guard.is_outdated = false;
-            core_guard.last_pinned = epoch;
-            core_guard.epoch_to_query_ids.insert(epoch, HashSet::new());
-        }
-        let last_pinned = core_guard.last_pinned;
-        core_guard
-            .epoch_to_query_ids
-            .get_mut(&last_pinned)
-            .unwrap()
-            .insert(query_id);
-        Ok(core_guard.last_pinned)
-    }
-
-    pub async fn unpin_snapshot(&self, epoch: u64, query_id: &QueryId) -> SchedulerResult<()> {
-        let min_epoch = async {
-            // Decrease the ref count of corresponding epoch. If all last pinned epoch is dropped,
-            // mark as outdated.
-            let mut core_guard = self.core.lock().await;
-            let query_ids = core_guard.epoch_to_query_ids.get_mut(&epoch);
-            if let Some(query_ids) = query_ids {
-                query_ids.remove(query_id);
-                if query_ids.is_empty() && epoch == core_guard.last_pinned {
-                    core_guard.is_outdated = true;
+        // do not use unbounded_channel because it may cause OOM when the RPC `get_epoch` blocks a
+        // long time.
+        let (sender, mut receiver) = channel(MAX_WAIT_EPOCH_REQUEST_NUM);
+        tokio::spawn(async move {
```

Comment:
Well, For FE, if we adopt push-based mechanism, we do not need this pull-based way.

We just need to update the cached epoch in notification service every time FE receive info from meta. It will simpler than current approach a lot. Described in #3577.

For unpin, we can simply design that we spwan a task to unpin every time receive a notification.  

The main concern for pull-based is it's epoch freshness. Might too stale compared to push.

Reply:
I want to KISS and be epoch-efficient in the early stage. What do you think?